Enabled feature contributions for GAM trainers

shmoradims opened the pull request:
* Implemented IFeatureContributionMapper interface for GamPredictorBase to enable FCC for GAM trainers
* Added test

artidoro commented on src/Microsoft.ML.FastTree/GamTrainer.cs:
```diff
@@ -1048,6 +1049,40 @@ public void SaveSummary(TextWriter writer, RoleMappedSchema schema)
+        }
+
+        private void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contributions,
+                        int top, int bottom, bool normalize)
```
Are you using the function from line 866: internal void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contribs, ref BufferBuilder<float> builder) 

If not, why not?

shmoradims replied:
Artidoro, I didn't see that one, but we should only keep one version of GetFeatureContributions. 

@rogancarr, In my version (line 1066) I'm using VBufferEditor and always use dense value of features, b/c the feature contributions could be significant even for features with value 0. So I didn't bother with dense/sparse checks, and the code is much shorter. I also have the top/bottom logic. I think you added the version on line 866, which uses the Intercept value and checks for dense/sparse.

Let's discuss which version to keep.

---
In reply to: [240429834](https://github.com/dotnet/machinelearning/pull/1856#discussion_r240429834) [](ancestors = 240429834)

rogancarr replied:
On the topic of multiplicity of functions, the `GetFeatureContributions` with this signature:
```cs
internal void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contribs, ref BufferBuilder<float> builder)
```

Is used by an internal tool. Let's leave it in place for now until we can remove it. I'm okay adding a separate method with the same name in that we use for `FeatureContributionCalculator` calls.

rogancarr replied:
Ignore the previous comment. You can get rid of the `BufferBuilder` implementation.

rogancarr replied:
On sparsity: I agree with this sentiment. It can be important that a feature is missing (zero).

rogancarr replied:
However, now that I've looked over the implementations, my suggestion would be to use the existing implementation, but
1. Change to using the `VBufferEditor`
2. Drop the `Intercept` term: This doesn't make sense
3. Change the method signature to match the new method signature.
4. Add the `SparsifyNormalize` call at the end.

The reason is that the sparsity handling of the existing function is nice because it doesn't densify the array to iterate over it and instead keeps track of the indices.

rogancarr replied:
As @shmoradims pointed out, `DenseValues()` is an iterator, so this should work as-is. Let's delete the old implementation then.

shmoradims replied:
Deleted the old code.

---
In reply to: [241154932](https://github.com/dotnet/machinelearning/pull/1856#discussion_r241154932) [](ancestors = 241154932)

Ivanidzo4ka commented:
>                 x4RandArray = new float[numberOfInstances];

That numFeatures variable looks bit out of place.
Any chance you want replace x1,x2,x3,x4 with two dimensional array of [numFeatures, numberOfinstances]?
Would be much easier to populate and iterate over.
(you can have randArray as well with  = [1000,10000, 5000, 1000]
and something like
```
for (i)
 for (j)
 arr[i,j] = rand.Next(rand[i]);
yArr[i] = arr[i,0]*10 + 20*arr[i,1]...
```

---
Refers to: test/Microsoft.ML.Tests/FeatureContributionTests.cs:81 in 1e1f0f3. [](commit_id = 1e1f0f33e7f660e1c901bbd23212c9578d7d996e, deletion_comment = False)

rogancarr commented on src/Microsoft.ML.FastTree/GamTrainer.cs:
```diff
@@ -649,7 +649,8 @@ public Stump(uint splitPoint, double lteValue, double gtValue)
     public abstract class GamPredictorBase : PredictorBase<float>,
-        IValueMapper, ICanSaveModel, ICanSaveInTextFormat, ICanSaveSummary
+        IValueMapper, ICanSaveModel, ICanSaveInTextFormat, ICanSaveSummary,
+        IFeatureContributionMapper
```
nit: Put after `IValueMapper`?

shmoradims replied:
done

---
In reply to: [240812855](https://github.com/dotnet/machinelearning/pull/1856#discussion_r240812855) [](ancestors = 240812855)

rogancarr commented on test/Microsoft.ML.Tests/FeatureContributionTests.cs:
```diff
@@ -30,15 +32,47 @@ public FeatureContributionTests(ITestOutputHelper output) : base(output)
         /// </summary>
-        [Fact]
-        public void TestFeatureImportance()
+        private void TestFeatureImportance(
```
`TestFeatureContribution`

shmoradims replied:
renamed

---
In reply to: [240855403](https://github.com/dotnet/machinelearning/pull/1856#discussion_r240855403) [](ancestors = 240855403)

shmoradims commented:
>                 x4RandArray = new float[numberOfInstances];

This was some old code, that I moved around. I refactored it as you mentioned. I had to use jagged array instead of 2d array, because, I wanted easy access to each row when creating the data view below.

---
In reply to: [446341134](https://github.com/dotnet/machinelearning/pull/1856#issuecomment-446341134) [](ancestors = 446341134)

---
Refers to: test/Microsoft.ML.Tests/FeatureContributionTests.cs:81 in 1e1f0f3. [](commit_id = 1e1f0f33e7f660e1c901bbd23212c9578d7d996e, deletion_comment = False)
